program/rust/src/rust_oracle.rs

```diff
@@ -227,6 +220,118 @@ pub fn add_mapping(
+        [x, y, _, z] => Ok([x, y, z]),
+        _ => Err(ProgramError::InvalidArgument),
+    }?;
+
+    check_valid_funding_account(funding_account)?;
+    check_valid_writable_account(program_id, price_account, size_of::<pc_price_t>())?;
+    // Check clock
+    let clock = Clock::from_account_info(clock_account)?;
+
+    let mut publisher_index: usize = 0;
+    let latest_aggregate_price: pc_price_info_t;
+    let latest_publisher_price: pc_price_info_t;
+    {
+        // Verify that symbol account is initialized
+        let price_data = load_checked::<pc_price_t>(price_account, cmd_args.ver_)?;
+
+        // Verify that publisher is authorized
+        while publisher_index < price_data.num_ as usize {
+            if pubkey_equal(
+                &price_data.comp_[publisher_index].pub_,
+                &funding_account.key.to_bytes(),
+            ) {
+                break;
+            }
+            publisher_index += 1;
+        }
+        pyth_assert(
+            publisher_index < price_data.num_ as usize,
+            ProgramError::InvalidArgument,
+        )?;
+
+
+        latest_publisher_price = price_data.comp_[publisher_index].latest_;
+        latest_aggregate_price = price_data.agg_;
+
+        // Check that publisher is publishing a more recent price
+        pyth_assert(
+            !is_component_update(cmd_args)?
+                || cmd_args.pub_slot_ > latest_publisher_price.pub_slot_,
+            ProgramError::InvalidArgument,
+        )?;
+    }
+
+    // Try to update the aggregate
+    let mut aggregate_updated = false;
+    if clock.slot > latest_aggregate_price.pub_slot_ {
+        unsafe {
+            aggregate_updated = c_upd_aggregate(
+                price_account.try_borrow_mut_data()?.as_mut_ptr(),
+                clock.slot,
+                clock.unix_timestamp,
+            );
+        }
+    }
+
+    // Try to update the publisher's price
+    if is_component_update(cmd_args)? {
+        let mut status: u32 = cmd_args.status_;
+        let mut threshold_conf = cmd_args.price_ / PC_MAX_CI_DIVISOR as i64;
+
+        if threshold_conf < 0 {
+            threshold_conf = -threshold_conf;
+        }
+
+        if cmd_args.conf_ > try_convert::<_, u64>(threshold_conf)? {
+            status = PC_STATUS_UNKNOWN
+        }
+
+        {
+            let mut price_data = load_checked::<pc_price_t>(price_account, cmd_args.ver_)?;
+            let publisher_price = &mut price_data.comp_[publisher_index].latest_;
+            publisher_price.price_ = cmd_args.price_;
+            publisher_price.conf_ = cmd_args.conf_;
+            publisher_price.status_ = status;
+            publisher_price.pub_slot_ = cmd_args.pub_slot_;
+        }
+    }
+
+    let account_len = price_account.try_data_len()?;
+    if aggregate_updated && account_len == PRICE_ACCOUNT_SIZE {
+        let mut price_account = load_account_as_mut::<PriceAccountWrapper>(price_account)?;
+        price_account.add_price_to_time_machine()?;
+    }
```

Comment:
nitpick: I think it would be better to put this block directly under the block that computes the new aggregate. (it's just easier to follow the code that way)